src/util/pointer_expr.h

```diff
@@ -1059,4 +1059,113 @@ inline void validate_expr(const object_size_exprt &value)
     "Object size expression must have pointer typed operand.");
 }
 
+/// A predicate that indicates that the object pointed to is live
+class live_object_exprt : public unary_predicate_exprt
+{
+public:
+  explicit live_object_exprt(exprt pointer)
+    : unary_predicate_exprt(ID_live_object, std::move(pointer))
+  {
+  }
+
+  exprt &pointer()
+  {
+    return op0();
+  }
+
+  const exprt &pointer() const
+  {
+    return op0();
+  }
+};
+
+template <>
+inline bool can_cast_expr<live_object_exprt>(const exprt &base)
+{
+  return base.id() == ID_live_object;
+}
+
+inline void validate_expr(const live_object_exprt &value)
+{
+  validate_operands(value, 1, "live_object must have one operand");
+}
+
+/// \brief Cast an exprt to a \ref live_object_exprt
+///
+/// \a expr must be known to be \ref live_object_exprt.
+///
+/// \param expr: Source expression
+/// \return Object of type \ref live_object_exprt
+inline const live_object_exprt &to_live_object_expr(const exprt &expr)
+{
+  PRECONDITION(expr.id() == ID_live_object);
+  const live_object_exprt &ret = static_cast<const live_object_exprt &>(expr);
+  validate_expr(ret);
+  return ret;
+}
+
+/// \copydoc to_live_object_expr(const exprt &)
+inline live_object_exprt &to_live_object_expr(exprt &expr)
+{
+  PRECONDITION(expr.id() == ID_live_object);
+  live_object_exprt &ret = static_cast<live_object_exprt &>(expr);
+  validate_expr(ret);
+  return ret;
+}
+
+/// A predicate that indicates that the object pointed to is writeable
+class writeable_object_exprt : public unary_predicate_exprt
```

Comment:
Does this imply a precondition that the pointer is valid? Should "pointer primitive checks" be expanded to cover this one? Should there be a check that the type of the argument passed to the constructor is a pointer?

Reply:
The pointer primitive checks could be expanded to cover this. However, since we are still sorting the user experience of writing contracts, this may be premature.